### PR TITLE
export dict of Click structure

### LINKED ISSUE #461: Export JSON Data of Click Structure
For external tools it would be nice if one could export the entire data of a click command into a JSON format.  For instance that way one could generate externally hosted documentation pages for commands that use better formatting than what `--help` generates.

### PR DESCRIPTION
**Pull request by YKo20010**
fixes #461

### CODE REVIEW

File: tests/test_as_dict.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
+import click
+
+
+def test_export_dict():
```
Comment (davidism) [inline tests/test_as_dict.py:4]: Split this into multiple tests for individual methods (`ParamType`, `Option`, `Command`, etc.).

**Comment by YKo20010** [comment]
@davidism We're ready for another review! Split up the tests - one for each param type, one for options, one for arguments, one for command, one for group, and one for nested groups.

**Comment by davidism** [comment]
Looks like some `test_` methods got added to the library code by mistake.

**Comment by YKo20010** [comment]
> Looks like some `test_` methods got added to the library code by mistake.
@davidism 
Are you talking about the `test_dict` methods in `core.py`? They're helper methods that compare the dictionaries from `to_info_dict` to an `expected` value. I originally wanted to put them within the test file `test_as_dict.py`, but then they were called as test cases when running `pytest`, which I didn't want, so I moved them to their respective classes.
What are your thoughts on where to put the `test_dict` helper methods?

**Comment by davidism** [comment]
Test code like that has to be in the tests. Renaming, or adding a `_` prefix, will work.

**Comment by davidism** [comment]
Can you squash this into one commit? Remember to add a `Co-authored-by` line.

**Comment by YKo20010** [comment]
@davidism Yeah, will do!

**Comment by davidism** [comment]
Looks good overall, pushed some cleanup.

* Removed methods that only called `super()`.
* `BaseCommand` takes `ctx` to match the subclasses.
* `Group` includes `chain`.
* `ParamType` includes a `param_type` key, which is the class name with the `ParamType` or `ParameterType` suffix removed. The `name` parameter is only a basic description, the type is needed to produce useful docs.
* Removed some keys from various objects. In general, considered what would be useful to user-facing docs vs what was an internal detail.

**Comment by davidism** [comment]
Made two changes to how `Group.commands` is handled. First, it needs to be a dict since the registered name might not match the command's name. Second, a new context is created to represent the subcommand, and is pushed so `get_current_context` will work.

Implemented `to_info_dict` on `Context`. The context itself can affect the CLI, and a project probably has some default context it calls the command with.

Added docs and changelog.

**Pull request merged**